Use expected for chunk::mmap

JoeLoser opened the pull request:
###  :notebook_with_decorative_cover: Description

<!-- Describe the change you've made in this section. -->

Problem:
- There are multiple ways that `chunk::mmap` can not succeed. In some
  cases, it logs and returns a `nullptr`, and in other cases, just returns
  a `nullptr`. In the former case, this can lead to duplicate logging
  since the callers of `mmap` already log on failure. However, the
  callers lack sufficient information as to exactly which of the calls
  inside `mmap` made it fail.

Solution:
- Change the return type of `chunk::map` to leverage `caf::expected` so
  we can carry the error message back to the callers so it is known
  exactly what failed when trying to perform the mmap operation.

###  :memo: Checklist

- [x] All user-facing changes have changelog entries.
- [x] The changes are reflected on [docs.tenzir.com/vast](https://docs.tenzir.com/vast), if necessary.
- [x] The PR description contains instructions for the reviewer, if necessary.

### :dart: Review Instructions

<!-- Provide instructions for the reviewer here, e.g., review this pull request commit-by-commit, or file-by-file. -->

File-by-file.

mavam commented:
You can rebase this PR now, @JoeLoser.

JoeLoser commented:
> You can rebase this PR now, @JoeLoser.

Rebased 👍

dominiklohmann commented on libvast/src/chunk.cpp:
```diff
@@ -53,26 +53,29 @@ chunk_ptr chunk::make(view_type view, deleter_type&& deleter) noexcept {
-      VAST_ERROR("failed to get file size for filename {}: {}", filename,
-                 err.message());
-      return nullptr;
+      return caf::make_error(ec::filesystem_error,
```
Nit: You can remove the surrounding braces per our style guide.

dominiklohmann commented on libvast/src/segment_store.cpp:
```diff
@@ -394,16 +394,17 @@ caf::error
   if (!chk)
-    return caf::make_error(ec::filesystem_error, "failed to mmap chunk",
-                           filename.string());
+    return caf::make_error(ec::filesystem_error,
+                           fmt::format("failed to mmap chunk {}: {}", filename,
+                                       caf::to_string(chk.error().context())));
```
It suffices to forward the error here.

```suggestion
    return std::move(chk.error());
```

dominiklohmann commented on libvast/src/segment_store.cpp:
```diff
@@ -427,9 +428,10 @@ caf::expected<segment> segment_store::load_segment(uuid id) const {
-    return caf::make_error(ec::filesystem_error, "failed to mmap chunk",
-                           filename.string());
-  if (auto segment = segment::make(std::move(chk))) {
+    return caf::make_error(ec::filesystem_error,
+                           fmt::format("failed to mmap chunk {}: {}", filename,
+                                       caf::to_string(chk.error().context())));
```
Same as above: It suffices to forward the error here.

```suggestion
    return std::move(chk.error());
```

dominiklohmann commented on libvast/src/system/index.cpp:
```diff
@@ -114,9 +114,11 @@ caf::error extract_partition_synopsis(
-    return caf::make_error(ec::system_error, "could not mmap partition at "
-                                               + partition_path.string());
-  const auto* partition = fbs::GetPartition(chunk->data());
+    return caf::make_error(
+      ec::filesystem_error,
+      fmt::format("could not mmap partition at {}: {}", partition_path,
+                  caf::to_string(chunk.error().context())));
```
```suggestion
    return std::move(chunk.error());
```

dominiklohmann commented on libvast/vast/chunk.hpp:
```diff
@@ -112,8 +112,9 @@ class chunk final : public caf::ref_counted {
   /// @param size The number of bytes to map. If 0, map the entire file.
   /// @param offset Where to start in terms of number of bytes from the start.
   /// @returns A chunk pointer or `nullptr` on failure.
```
```suggestion
  /// @returns A chunk pointer or an error on failure. The returned chunk
  /// pointer is never `nullptr`.
```

dominiklohmann reviewed (changes requested):
Just some very minor requests, other than that this looks really good. I double-checked that you caught all usages of `chunk::mmap`.

JoeLoser commented:
> Just some very minor requests, other than that this looks really good. I double-checked that you caught all usages of `chunk::mmap`.

Thanks for the review. I believe I've addressed all of the feedback comments; please re-review.
